crates/evm-ethereum/src/execute.rs

```diff
@@ -0,0 +1,826 @@
+    };
+    use reth_revm::{
+        database::StateProviderDatabase, test_utils::StateProviderTest, TransitionState,
+    };
+    use std::collections::HashMap;
+
+    static BEACON_ROOT_CONTRACT_CODE: Bytes = bytes!("3373fffffffffffffffffffffffffffffffffffffffe14604d57602036146024575f5ffd5b5f35801560495762001fff810690815414603c575f5ffd5b62001fff01545f5260205ff35b5f5ffd5b62001fff42064281555f359062001fff015500");
+
+    fn create_state_provider_with_beacon_root_contract() -> StateProviderTest {
+        let mut db = StateProviderTest::default();
+
+        let beacon_root_contract_account = Account {
+            balance: U256::ZERO,
+            bytecode_hash: Some(keccak256(BEACON_ROOT_CONTRACT_CODE.clone())),
+            nonce: 1,
+        };
+
+        db.insert_account(
+            BEACON_ROOTS_ADDRESS,
+            beacon_root_contract_account,
+            Some(BEACON_ROOT_CONTRACT_CODE.clone()),
+            HashMap::new(),
+        );
+
+        db
+    }
+
+    fn executor_provider(chain_spec: Arc<ChainSpec>) -> EthExecutorProvider<EthEvmConfig> {
+        EthExecutorProvider {
+            chain_spec,
+            evm_config: Default::default(),
+            inspector: None,
+            prune_modes: Default::default(),
+        }
+    }
+
+    #[test]
+    fn eip_4788_non_genesis_call() {
+        let mut header =
+            Header { timestamp: 1, number: 1, excess_blob_gas: Some(0), ..Header::default() };
+
+        let db = create_state_provider_with_beacon_root_contract();
+
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::from(&*MAINNET)
+                .shanghai_activated()
+                .with_fork(Hardfork::Cancun, ForkCondition::Timestamp(1))
+                .build(),
+        );
+
+        let provider = executor_provider(chain_spec);
+
+        // attempt to execute a block without parent beacon block root, expect err
+        let err = provider
+            .executor(StateProviderDatabase::new(&db))
+            .execute(
+                (
+                    &BlockWithSenders {
+                        block: Block {
+                            header: header.clone(),
+                            body: vec![],
+                            ommers: vec![],
+                            withdrawals: None,
+                        },
+                        senders: vec![],
+                    },
+                    U256::ZERO,
+                )
+                    .into(),
+            )
+            .expect_err(
+                "Executing cancun block without parent beacon block root field should fail",
+            );
+        assert_eq!(
+            err,
+            BlockExecutionError::Validation(BlockValidationError::MissingParentBeaconBlockRoot)
+        );
+
+        // fix header, set a gas limit
+        header.parent_beacon_block_root = Some(B256::with_last_byte(0x69));
+
+        let mut executor = provider.executor(StateProviderDatabase::new(&db));
```

Comment:
What about an API along the lines of:

`ExecutorProvider::builder().with_chain_spec(chain_spec).with_db(db)` where `with_db<DB: Into<StateProviderDatabase>>`?